torch/csrc/autograd/engine.cpp

```diff
@@ -885,33 +985,54 @@ bool Engine::is_checkpoint_valid() {
   return checkpoint_valid;
 }
 
-size_t Engine::ready_queue_size(at::Device device) {
-  if (ready_queues_.empty()) {
+std::shared_ptr<ReadyQueue> Engine::init_local_ready_queue(std::shared_ptr<ReadyQueue> ready_queue) {
+  // if provided ready queue in the caller, use the caller's ready_queue to initialize local_ready_queue
+  if (ready_queue) {
+    TORCH_INTERNAL_ASSERT(!local_ready_queue);
+    local_ready_queue = std::move(ready_queue);
+  } else if (!local_ready_queue){
+    // otherwise if local_ready_queue not allocated, allocate a new ready_queue
+    local_ready_queue = std::make_shared<ReadyQueue>();
+  }
+  return local_ready_queue;
+}
+
+size_t Engine::ready_queue_size(const std::shared_ptr<GraphTask>& graph_task, at::Device device) {
+  if (device_ready_queues_.empty()) {
     // The vector ready_queues_ is initialized in start_threads, but this method
     // can be called before start_threads. Adding this check to avoid index
     // out of bound error.
     return 0;
   }
-  return ready_queue(device).size();
+  return ready_queue(graph_task, device).size();
 }
 
-auto Engine::ready_queue(at::Device device) -> ReadyQueue& {
-  // See Note [Allocating GPUs to autograd threads]
+// CPU ready queue is per GraphTask, but CUDA device ready queues are shared across all graph tasks
+auto Engine::ready_queue(const std::shared_ptr<GraphTask>& graph_task, at::Device device) -> ReadyQueue& {
```

Comment:
You can remove duplicate code by just having this function call `ready_queue_by_index` with `device.index()` no?

Reply:
double checked [here](https://github.com/pytorch/pytorch/blob/33e7977154a6ca961ce927e15625763f9cc13f8c/c10/core/DeviceType.h#L15), `device.index()` in CPU returns 0, but we would like the cpu device index be `-1` to distinguish with `cuda:0`, also `device.index()` only have `cuda` instead of different cuda hosts. so this two are different semantic and usage. `ready_queue(at::Device)` is used in thread_main to push cross device training tasks. `ready_queue_by_index(int)` is mainly used to notify the owner device not sleeping on pop.